### PR TITLE
sql: add cluster version gating for new schema changer

### LINKED ISSUE #77151: clusterversion: declarative schema changer version gate
~We should push it down past the descriptor migration which adds IDs to constraints.~

We should add it at the end of the `Key` enum in `cockroach_versions.go`. This gate should be exercised at the `scbuild.Build` call site, or perhaps further down the call stack, I can't tell.

Jira issue: CRDB-13436

### PR DESCRIPTION
**Pull request by chengxiong-ruan**
resolves: #77151

Release note: None
Release justification: version gate is needed for the release.

### CODE REVIEW

File: pkg/sql/opaque.go [hunk 1]
```diff
@@ -47,7 +48,9 @@ func buildOpaque(
 	scalarProps.Require(stmt.StatementTag(), tree.RejectSubqueries)
 
 	var plan planNode
-	if tree.CanModifySchema(stmt) {
+	if evalCtx.Settings.Version.IsActive(ctx, clusterversion.EnableDeclarativeSchemaChanger) &&
+		tree.CanModifySchema(stmt) {
+
 		if err := p.checkNoConflictingCursors(stmt); err != nil {
```
Comment (postamar) [inline pkg/sql/opaque.go:54]: I believe this check needs to be performed regardless of the version gate.

**Review by postamar — CHANGES_REQUESTED** [review]
You probably should test this somehow. An unupgraded cluster at version `EnableDeclarativeSchemaChanger-1` should not be able to perform `EXPLAIN (DDL) DROP DATABASE defaultdb`.

**Comment by chengxiong-ruan** [comment]
___
*[pkg/sql/opaque.go, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77242#-Mx61cpCMdNuIkYTeEXD-r1-54:-Mx6PspkE8vg8WAX45pU:bswoy71) ([raw file](https://github.com/cockroachdb/cockroach/blob/262ca917a9a45be01759f1b733a88c148e296fea/pkg/sql/opaque.go#L54)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I believe this check needs to be performed regardless of the version gate.
</blockquote></details>

oh man, sorry :(

**Review by chengxiong-ruan — COMMENTED** [review]
Thanks, test was added, I was confused by `fully supported` and didn't realize `alter table` has a second level of `fully supported` on each commands haha...so played with it for a while >_>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77242#-:-Mx6hRX_73ik-JksF3el:b-cb2bf8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar)

___
*[pkg/sql/opaque.go, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77242#-Mx61cpCMdNuIkYTeEXD-r1-54:-Mx6hjsb7PxfCGQdeQiS:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/262ca917a9a45be01759f1b733a88c148e296fea/pkg/sql/opaque.go#L54)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

oh man, sorry :(
</blockquote></details>

Fixed

**Review by postamar — APPROVED** [review]
:lgtm: 

> Reviewed 4 of 5 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77242#-:-MxAG7JpBHLl7AoHOgZ5:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @chengxiong-ruan)

___
*[pkg/sql/schemachanger/schemachanger_test.go, line 624 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77242#-MxAFKTQDY4EVN67BKro:-MxAFKTQDY4EVN67BKrp:b-ccjln6) ([raw file](https://github.com/cockroachdb/cockroach/blob/20ba70b0481b27bcd300e62508e852c7589bcb0a/pkg/sql/schemachanger/schemachanger_test.go#L624)):*
> ```Go
> }
> 
> func TestNewSchemaChangerVersionGating(t *testing.T) {
> ```

nit: add a TODO that states that this test can be removed from `master` once v22.1 has been released.
___
*[pkg/sql/schemachanger/schemachanger_test.go, line 671 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77242#-MxAFa__3VUv2svuper1:-MxAFa__3VUv2svuper2:b-t960o2) ([raw file](https://github.com/cockroachdb/cockroach/blob/20ba70b0481b27bcd300e62508e852c7589bcb0a/pkg/sql/schemachanger/schemachanger_test.go#L671)):*
> ```Go
> 		)
> 	})
> }
> ```

I'm fine with this. If you're super paranoid you can also add a case where you `DROP DATABASE db CASCADE` and check that no jobs exist of type`NEW SCHEMA CHANGE` when the declarative schema changer is disabled, and vice versa..

**Review by chengxiong-ruan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77242)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @postamar)

___
*[pkg/sql/schemachanger/schemachanger_test.go, line 624 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77242#-MxAFKTQDY4EVN67BKro:-MxAd5cw1N5rnzAicrUZ:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/20ba70b0481b27bcd300e62508e852c7589bcb0a/pkg/sql/schemachanger/schemachanger_test.go#L624)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: add a TODO that states that this test can be removed from `master` once v22.1 has been released.
</blockquote></details>

done
___
*[pkg/sql/schemachanger/schemachanger_test.go, line 671 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77242#-MxAFa__3VUv2svuper1:-MxAd6gI6DQiVCnbFDdN:b-80kasa) ([raw file](https://github.com/cockroachdb/cockroach/blob/20ba70b0481b27bcd300e62508e852c7589bcb0a/pkg/sql/schemachanger/schemachanger_test.go#L671)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I'm fine with this. If you're super paranoid you can also add a case where you `DROP DATABASE db CASCADE` and check that no jobs exist of type`NEW SCHEMA CHANGE` when the declarative schema changer is disabled, and vice versa..
</blockquote></details>

thanks for tip, and I'm not paranoid :)

**Comment by chengxiong-ruan** [comment]
TFTR!
bors r+

**Pull request merged**